dxr/plugins/extmatch.py

```diff
@@ -0,0 +1,73 @@
+"""Extmatch - Match files in a given folder by extension and provide a link
+in the nav panel of each to the other.
+
+Currently we match "Header" file extensions with "Implementation" file
+extensions.
+"""
+
+from fnmatch import fnmatchcase
+from os.path import relpath, join, basename, splitext, isfile
+from collections import namedtuple
+
+from dxr.indexers import (FileToIndex as FileToIndexBase,
+                          TreeToIndex as TreeToIndexBase)
+from dxr.mime import icon
+from dxr.utils import browse_file_url
+
+# A tuple of extensions along with a title describing the type of extensions:
+_TitledExts = namedtuple('_TitledExts', ['exts', 'title'])
+
+
+class TreeToIndex(TreeToIndexBase):
+    def __init__(self, plugin_name, tree, vcs_cache):
+        super(TreeToIndex, self).__init__(plugin_name, tree, vcs_cache)
+        self.header_exts = _TitledExts(('.h', '.hpp', '.hxx'), 'Header')
```

Comment:
> Quite possibly we should use something more efficient instead, since we're doing a lot of linear scanning over them using in. But the data sizes are small, so we might not see a benefit.

This:
```python
impl_exts_list = ['.cpp', '.c', '.cc', '.cxx', '.mm']
h_exts_list = ['.h', '.hpp', '.hxx']

impl_exts_set = set(['.cpp', '.c', '.cc', '.cxx', '.mm'])
h_exts_set = set(['.h', '.hpp', '.hxx'])

for impl_exts, h_exts in [[impl_exts_list, h_exts_list],
                          [impl_exts_set, h_exts_set],
                          [impl_exts_list, h_exts_list],
                          [impl_exts_set, h_exts_set]]:
    
    start = timer()
    for dirpath, dirnames, filenames in os.walk('./mozilla-central'):
        for f in filenames:
            root, ext = splitext(f)
            if ext in impl_exts:
                exts = h_exts
            if ext in h_exts:
                exts = impl_exts
    print timer() - start
```
prints
```
4.68175315857
4.64918088913
4.69052195549
4.64750003815
```
(once the m-c listings are in memory), so in this case it indeed looks like it doesn't make much difference.  The other issue with using `set` in this case is that when we iterate over the matched (dual) extensions, we want to do the `is_file` check in the order the extensions were listed since they're in most-frequent-extension-first order, which (theoretically) need not happen if we iterate sets instead.  Or did you have another container in mind?  If not I think I'll just switch to lists.

(Thanks for bringing this up in any case :-)